File: deedlit.dev.comfyhelper/lib/metadata-parsing.ts
export type PromptMetadataInsights = {
  positivePrompt?: string;
  negativePrompt?: string;
  model?: string;
  sampler?: string;
  cfgScale?: string;
  steps?: string;
  seed?: string;
  scheduler?: string;
  size?: string;
};

const metadataSearchCache = new WeakMap<object, Map<string, unknown>>();

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeKey(key: string): string {
  return key.toLowerCase().replace(/[^a-z0-9]/g, "");
}

function getCacheKey(keys: string[]): string {
  return keys.map((key) => normalizeKey(key)).join("|");
}

export function maybeParseJsonString(value: unknown): unknown {
  if (typeof value !== "string") {
    return value;
  }

  const trimmed = value.trim();
  if (!trimmed || !(trimmed.startsWith("{") || trimmed.startsWith("["))) {
    return value;
  }

  try {
    return JSON.parse(trimmed) as unknown;
  } catch {
    return value;
  }
}

export function toDisplayValue(value: unknown): string | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }

  if (typeof value === "string") {
    const trimmed = value.trim();
    return trimmed ? trimmed : undefined;
  }

  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }

  return undefined;
}

export function getSearchableMetadata(metadata: unknown): unknown {
  return isRecord(metadata) && isRecord(metadata.fields) ? metadata.fields : metadata;
}

export function findFirstValueByKeys(root: unknown, keys: string[]): unknown {
  if (!root) {
    return undefined;
  }

  if (isRecord(root)) {
    const cache = metadataSearchCache.get(root);
    const cacheKey = getCacheKey(keys);
    if (cache?.has(cacheKey)) {
      return cache.get(cacheKey);
    }
  }

  const keySet = new Set(keys.map((key) => normalizeKey(key)));
  const seen = new WeakSet<object>();

  const visit = (value: unknown): unknown => {
    const parsed = maybeParseJsonString(value);

    if (Array.isArray(parsed)) {
      for (const item of parsed) {
        const found = visit(item);
        if (found !== undefined) {
          return found;
        }
      }
      return undefined;
    }

    if (!isRecord(parsed)) {
      return undefined;
    }

    if (seen.has(parsed)) {
      return undefined;
    }
    seen.add(parsed);

    for (const [key, nestedValue] of Object.entries(parsed)) {
      if (keySet.has(normalizeKey(key))) {
        return nestedValue;
      }
    }

    for (const nestedValue of Object.values(parsed)) {
      const found = visit(nestedValue);
      if (found !== undefined) {
        return found;
      }
    }

    return undefined;
  };

  const result = visit(root);

  if (isRecord(root)) {
    const cacheKey = getCacheKey(keys);
    const cache = metadataSearchCache.get(root) ?? new Map<string, unknown>();
    cache.set(cacheKey, result);
    metadataSearchCache.set(root, cache);
  }

  return result;
}

export function extractPromptTextFromMetadata(metadata: unknown, keys: string[]): string | undefined {
  return toDisplayValue(findFirstValueByKeys(metadata, keys));
}

function resolveNodeReference(value: unknown): string | null {
  if (Array.isArray(value) && value.length > 0) {
    const first = value[0];
    if (typeof first === "string" || typeof first === "number") {
      return String(first);
    }
  }

  if (typeof value === "string" || typeof value === "number") {
    return String(value);
  }

  return null;
}

function extractTextFromPromptNode(
  nodeId: string | null,
  nodes: Record<string, { classType?: string; inputs?: Record<string, unknown> }>,
): string | undefined {
  if (!nodeId) {
    return undefined;
  }

  const node = nodes[nodeId];
  if (!node?.inputs) {
    return undefined;
  }

  const parts = [
    toDisplayValue(node.inputs.text),
    toDisplayValue(node.inputs.text_g),
    toDisplayValue(node.inputs.text_l),
  ].filter((value): value is string => Boolean(value));

  if (parts.length === 0) {
    return undefined;
  }

  return parts.join("\n");
}

export function extractFromComfyPromptGraph(promptValue: unknown): Partial<PromptMetadataInsights> {
  const parsedPrompt = maybeParseJsonString(promptValue);
  if (!isRecord(parsedPrompt)) {
    return {};
  }

  const nodes: Record<string, { classType?: string; inputs?: Record<string, unknown> }> = {};

  for (const [nodeId, nodeValue] of Object.entries(parsedPrompt)) {
    if (!isRecord(nodeValue)) {
      continue;
    }

    const classType = toDisplayValue(nodeValue.class_type);
    const inputs = isRecord(nodeValue.inputs) ? nodeValue.inputs : undefined;
    if (!classType || !inputs) {
      continue;
    }

    nodes[nodeId] = { classType, inputs };
  }

  let kSamplerEntry: [string, { classType?: string; inputs?: Record<string, unknown> }] | undefined;
  let loaderNode: { classType?: string; inputs?: Record<string, unknown> } | undefined;

  for (const [nodeId, node] of Object.entries(nodes)) {
    const classType = (node.classType ?? "").toLowerCase();

    if (!kSamplerEntry && classType.includes("ksampler")) {
      kSamplerEntry = [nodeId, node];
    }

    if (!loaderNode && (classType.includes("checkpointloader") || classType.includes("unetloader"))) {
      loaderNode = node;
    }

    if (kSamplerEntry && loaderNode) {
      break;
    }
  }

  if (!kSamplerEntry) {
    return {};
  }

  const [, samplerNode] = kSamplerEntry;
  const inputs = samplerNode.inputs ?? {};
  const positiveRef = resolveNodeReference(inputs.positive);
  const negativeRef = resolveNodeReference(inputs.negative);
  const modelRef = resolveNodeReference(inputs.model);

  let modelName: string | undefined;
  if (modelRef && nodes[modelRef]?.inputs) {
    const modelInputs = nodes[modelRef].inputs ?? {};
    modelName =
      toDisplayValue(modelInputs.ckpt_name) ??
      toDisplayValue(modelInputs.model_name) ??
      toDisplayValue(modelInputs.unet_name);
  }

  if (!modelName && !loaderNode) {
    loaderNode = Object.values(nodes).find((node) => {
      const classType = (node.classType ?? "").toLowerCase();
      return classType.includes("checkpointloader") || classType.includes("unetloader");
    });
  }

  if (!modelName && loaderNode?.inputs) {
    modelName =
      toDisplayValue(loaderNode.inputs.ckpt_name) ??
      toDisplayValue(loaderNode.inputs.model_name) ??
      toDisplayValue(loaderNode.inputs.unet_name);
  }

  return {
    positivePrompt: extractTextFromPromptNode(positiveRef, nodes),
    negativePrompt: extractTextFromPromptNode(negativeRef, nodes),
    model: modelName,
    sampler: toDisplayValue(inputs.sampler_name),
    cfgScale: toDisplayValue(inputs.cfg),
    steps: toDisplayValue(inputs.steps),
    seed: toDisplayValue(inputs.seed),
    scheduler: toDisplayValue(inputs.scheduler),
  };
}

type ParseAutomatic1111ParametersOptions = {
  includeFirstLineAsPositive?: boolean;
};

export function parseAutomatic1111Parameters(
  parameters: string,
  options: ParseAutomatic1111ParametersOptions = {},
): Partial<PromptMetadataInsights> {
  const result: Partial<PromptMetadataInsights> = {};
  const text = parameters.trim();

  if (!text) {
    return result;
  }

  const negativeLabel = "Negative prompt:";
  const negativeIndex = text.indexOf(negativeLabel);
  if (negativeIndex >= 0) {
    const positivePart = text.slice(0, negativeIndex).trim();
    if (positivePart) {
      result.positivePrompt = positivePart;
    }

    const afterNegative = text.slice(negativeIndex + negativeLabel.length);
    const settingsStart = afterNegative.search(/\n(?:Steps|Sampler|CFG scale|Seed|Size|Model):/i);
    const negativePart = settingsStart >= 0 ? afterNegative.slice(0, settingsStart).trim() : afterNegative.trim();
    if (negativePart) {
      result.negativePrompt = negativePart;
    }
  } else if (options.includeFirstLineAsPositive) {
    const firstLine = text.split("\n")[0]?.trim();
    if (firstLine) {
      result.positivePrompt = firstLine;
    }
  }

  const capture = (pattern: RegExp): string | undefined => text.match(pattern)?.[1]?.trim();

  result.steps = capture(/Steps:\s*([^,\n]+)/i);
  result.sampler = capture(/Sampler:\s*([^,\n]+)/i);
  result.cfgScale = capture(/CFG scale:\s*([^,\n]+)/i);
  result.seed = capture(/Seed:\s*([^,\n]+)/i);
  result.size = capture(/Size:\s*([^,\n]+)/i);
  result.model = capture(/Model:\s*([^,\n]+)/i);

  return result;
}

File: deedlit.metadata/metadata_parsing.py
"""Metadata interpretation helpers.

Ported faithfully from ``lib/metadata-parsing.ts``:
- recursive key search (``findFirstValueByKeys``) with key normalization
- JSON-string coercion (``maybeParseJsonString``)
- A1111 ``parameters`` string parsing (``parseAutomatic1111Parameters``)
- ComfyUI prompt-graph walk (``extractFromComfyPromptGraph``)

The ComfyUI walk is intentionally structured so #7 (references) can extend the
node-graph traversal here to resolve checkpoints/loras/embeddings/etc. without
reworking the prompt/param extraction.
"""
from __future__ import annotations

import json
import re
from typing import Any

__all__ = [
    "is_record",
    "maybe_parse_json_string",
    "to_display_value",
    "get_searchable_metadata",
    "find_first_value_by_keys",
    "extract_prompt_text_from_metadata",
    "parse_automatic1111_parameters",
    "extract_from_comfy_prompt_graph",
]

_NON_ALNUM_RE = re.compile(r"[^a-z0-9]")


def is_record(value: Any) -> bool:
    """TS ``isRecord``: plain object (dict), not a list."""
    return isinstance(value, dict)


def _normalize_key(key: str) -> str:
    return _NON_ALNUM_RE.sub("", key.lower())


def maybe_parse_json_string(value: Any) -> Any:
    if not isinstance(value, str):
        return value
    trimmed = value.strip()
    if not trimmed or not (trimmed.startswith("{") or trimmed.startswith("[")):
        return value
    try:
        return json.loads(trimmed)
    except json.JSONDecodeError:
        return value


def to_display_value(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, str):
        trimmed = value.strip()
        return trimmed if trimmed else None
    # NOTE: bool is a subclass of int in Python; check it explicitly. TS renders
    # booleans as "true"/"false"; Python's str(True) is "True". This only ever
    # affects boolean inputs (none of the params we read are booleans).
    if isinstance(value, (int, float)):
        return str(value)
    return None


def get_searchable_metadata(metadata: Any) -> Any:
    if is_record(metadata) and is_record(metadata.get("fields")):
        return metadata["fields"]
    return metadata


def find_first_value_by_keys(root: Any, keys: list[str]) -> Any:
    """Recursive first-match search over the metadata tree.

    Mirrors the TS DFS: at each record, return the value of the first own key
    whose normalized form is in ``keys``; otherwise descend into values. JSON
    strings are coerced on the way down. Cycles are guarded by identity.
    """
    if not root:
        return None

    key_set = {_normalize_key(k) for k in keys}
    seen: set[int] = set()

    def visit(value: Any) -> Any:
        parsed = maybe_parse_json_string(value)

        if isinstance(parsed, list):
            for item in parsed:
                found = visit(item)
                if found is not None:
                    return found
            return None

        if not is_record(parsed):
            return None

        ident = id(parsed)
        if ident in seen:
            return None
        seen.add(ident)

        for key, nested in parsed.items():
            if _normalize_key(key) in key_set:
                return nested

        for nested in parsed.values():
            found = visit(nested)
            if found is not None:
                return found

        return None

    return visit(root)


def extract_prompt_text_from_metadata(metadata: Any, keys: list[str]) -> str | None:
    return to_display_value(find_first_value_by_keys(metadata, keys))


# ---------------------------------------------------------------------------
# A1111 `parameters`
# ---------------------------------------------------------------------------

_SETTINGS_START_RE = re.compile(r"\n(?:Steps|Sampler|CFG scale|Seed|Size|Model):", re.IGNORECASE)


def _capture(text: str, pattern: re.Pattern[str]) -> str | None:
    m = pattern.search(text)
    if not m:
        return None
    return m.group(1).strip()


_STEPS_RE = re.compile(r"Steps:\s*([^,\n]+)", re.IGNORECASE)
_SAMPLER_RE = re.compile(r"Sampler:\s*([^,\n]+)", re.IGNORECASE)
_CFG_RE = re.compile(r"CFG scale:\s*([^,\n]+)", re.IGNORECASE)
_SEED_RE = re.compile(r"Seed:\s*([^,\n]+)", re.IGNORECASE)
_SIZE_RE = re.compile(r"Size:\s*([^,\n]+)", re.IGNORECASE)
_MODEL_RE = re.compile(r"Model:\s*([^,\n]+)", re.IGNORECASE)


def parse_automatic1111_parameters(
    parameters: str, *, include_first_line_as_positive: bool = False
) -> dict[str, str]:
    """Port of ``parseAutomatic1111Parameters``.

    Returns only the keys that were found (matching the TS ``Partial`` shape):
    ``positivePrompt``, ``negativePrompt``, ``steps``, ``sampler``,
    ``cfgScale``, ``seed``, ``size``, ``model``.
    """
    result: dict[str, str] = {}
    text = parameters.strip()
    if not text:
        return result

    negative_label = "Negative prompt:"
    negative_index = text.find(negative_label)
    if negative_index >= 0:
        positive_part = text[:negative_index].strip()
        if positive_part:
            result["positivePrompt"] = positive_part

        after_negative = text[negative_index + len(negative_label) :]
        m = _SETTINGS_START_RE.search(after_negative)
        settings_start = m.start() if m else -1
        negative_part = (
            after_negative[:settings_start].strip()
            if settings_start >= 0
            else after_negative.strip()
        )
        if negative_part:
            result["negativePrompt"] = negative_part
    elif include_first_line_as_positive:
        first_line = text.split("\n")[0].strip()
        if first_line:
            result["positivePrompt"] = first_line

    for name, pattern in (
        ("steps", _STEPS_RE),
        ("sampler", _SAMPLER_RE),
        ("cfgScale", _CFG_RE),
        ("seed", _SEED_RE),
        ("size", _SIZE_RE),
        ("model", _MODEL_RE),
    ):
        captured = _capture(text, pattern)
        if captured is not None:
            result[name] = captured

    return result


# ---------------------------------------------------------------------------
# ComfyUI graph walk
# ---------------------------------------------------------------------------


def _resolve_node_reference(value: Any) -> str | None:
    if isinstance(value, list) and len(value) > 0:
        first = value[0]
        if isinstance(first, (str, int, float)) and not isinstance(first, bool):
            return str(first)
        if isinstance(first, str):
            return first
    if isinstance(value, (str, int, float)) and not isinstance(value, bool):
        return str(value)
    return None


def _extract_text_from_prompt_node(node_id: str | None, nodes: dict[str, dict]) -> str | None:
    if not node_id:
        return None
    node = nodes.get(node_id)
    if not node or not node.get("inputs"):
        return None
    inputs = node["inputs"]
    parts = [
        to_display_value(inputs.get("text")),
        to_display_value(inputs.get("text_g")),
        to_display_value(inputs.get("text_l")),
    ]
    parts = [p for p in parts if p]
    if not parts:
        return None
    return "\n".join(parts)


def extract_from_comfy_prompt_graph(prompt_value: Any) -> dict[str, str | None]:
    """Port of ``extractFromComfyPromptGraph``.

    Walks the ComfyUI api-prompt node graph: find the first KSampler-like node
    and a checkpoint/unet loader, then resolve positive/negative prompt text,
    model name, and sampler params. Returns a dict of the keys that were found
    (TS ``Partial`` shape).
    """
    parsed_prompt = maybe_parse_json_string(prompt_value)
    if not is_record(parsed_prompt):
        return {}

    nodes: dict[str, dict] = {}
    for node_id, node_value in parsed_prompt.items():
        if not is_record(node_value):
            continue
        class_type = to_display_value(node_value.get("class_type"))
        inputs = node_value.get("inputs") if is_record(node_value.get("inputs")) else None
        if not class_type or inputs is None:
            continue
        nodes[node_id] = {"classType": class_type, "inputs": inputs}

    ksampler_entry: tuple[str, dict] | None = None
    loader_node: dict | None = None

    for node_id, node in nodes.items():
        class_type = (node.get("classType") or "").lower()
        if ksampler_entry is None and "ksampler" in class_type:
            ksampler_entry = (node_id, node)
        if loader_node is None and (
            "checkpointloader" in class_type or "unetloader" in class_type
        ):
            loader_node = node
        if ksampler_entry and loader_node:
            break

    if ksampler_entry is None:
        return {}

    _, sampler_node = ksampler_entry
    inputs = sampler_node.get("inputs") or {}
    positive_ref = _resolve_node_reference(inputs.get("positive"))
    negative_ref = _resolve_node_reference(inputs.get("negative"))
    model_ref = _resolve_node_reference(inputs.get("model"))

    model_name: str | None = None
    if model_ref and nodes.get(model_ref, {}).get("inputs"):
        model_inputs = nodes[model_ref]["inputs"] or {}
        model_name = (
            to_display_value(model_inputs.get("ckpt_name"))
            or to_display_value(model_inputs.get("model_name"))
            or to_display_value(model_inputs.get("unet_name"))
        )

    if not model_name and not loader_node:
        for node in nodes.values():
            class_type = (node.get("classType") or "").lower()
            if "checkpointloader" in class_type or "unetloader" in class_type:
                loader_node = node
                break

    if not model_name and loader_node and loader_node.get("inputs"):
        loader_inputs = loader_node["inputs"]
        model_name = (
            to_display_value(loader_inputs.get("ckpt_name"))
            or to_display_value(loader_inputs.get("model_name"))
            or to_display_value(loader_inputs.get("unet_name"))
        )

    return {
        "positivePrompt": _extract_text_from_prompt_node(positive_ref, nodes),
        "negativePrompt": _extract_text_from_prompt_node(negative_ref, nodes),
        "model": model_name,
        "sampler": to_display_value(inputs.get("sampler_name")),
        "cfgScale": to_display_value(inputs.get("cfg")),
        "steps": to_display_value(inputs.get("steps")),
        "seed": to_display_value(inputs.get("seed")),
        "scheduler": to_display_value(inputs.get("scheduler")),
    }
